File: richeditor/src/main/java/jp/wasabeef/richeditor/RichEditor.java
package jp.wasabeef.richeditor;

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Bitmap;
import android.graphics.drawable.Drawable;
import android.net.Uri;
import android.os.Build;
import android.text.TextUtils;
import android.util.AttributeSet;
import android.util.Log;
import android.view.Gravity;
import android.webkit.WebChromeClient;
import android.webkit.WebView;
import android.webkit.WebViewClient;

import java.io.UnsupportedEncodingException;
import java.net.URLDecoder;
import java.net.URLEncoder;
import java.util.HashMap;
import java.util.List;

/**
 * Copyright (C) 2017 Wasabeef
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

public class RichEditor extends WebView {

  public class EnabledFormatTypes {
    private boolean isBold;
    private boolean isItalic;
    private boolean isUnderline;
    private boolean isStrikethrough;
    private boolean isUnorderedList;
    private boolean isOrderedList;
    private boolean isSuperscript;
    private boolean isSubscript;

    public EnabledFormatTypes(boolean isBold, boolean isItalic, boolean isUnderline, boolean isStrikethrough, boolean isUnorderedList, boolean isOrderedList, boolean isSuperscript, boolean isSubscript) {
      this.isBold = isBold;
      this.isItalic = isItalic;
      this.isUnderline = isUnderline;
      this.isStrikethrough = isStrikethrough;
      this.isUnorderedList = isUnorderedList;
      this.isOrderedList = isOrderedList;
      this.isSuperscript = isSuperscript;
      this.isSubscript = isSubscript;
    }

    public boolean isBold() {
      return isBold;
    }
    public boolean isItalic() {
      return isItalic;
    }
    public boolean isUnderline() {
      return isUnderline;
    }
    public boolean isStrikethrough() {
      return isStrikethrough;
    }
    public boolean isUnorderedList() { return isUnorderedList; }
    public boolean isOrderedList() { return isOrderedList; }
    public boolean isSuperscript() {
      return isSuperscript;
    }
    public boolean isSubscript() {
      return isSubscript;
    }

    public HashMap<String, Boolean> getEnabledTypesOnly() {
      HashMap<String, Boolean> enabledMap = new HashMap<>();
      if (isBold)          enabledMap.put("isBold", isBold);
      if (isItalic)        enabledMap.put("isItalic", isItalic);
      if (isUnderline)     enabledMap.put("isUnderline", isUnderline);
      if (isStrikethrough) enabledMap.put("isStrikethrough", isStrikethrough);
      if (isUnorderedList) enabledMap.put("isUnorderedList", isUnorderedList);
      if (isOrderedList)   enabledMap.put("isOrderedList", isOrderedList);
      if (isSuperscript)   enabledMap.put("isSuperscript", isSuperscript);
      if (isSubscript)     enabledMap.put("isSubscript", isSubscript);
      return enabledMap;
    }

    public HashMap<String, Boolean> getAllTypes() {
      HashMap<String, Boolean> enabledMap = new HashMap<>();
      enabledMap.put("isBold", isBold);
      enabledMap.put("isItalic", isItalic);
      enabledMap.put("isUnderline", isUnderline);
      enabledMap.put("isStrikethrough", isStrikethrough);
      enabledMap.put("isUnorderedList", isUnorderedList);
      enabledMap.put("isOrderedList", isOrderedList);
      enabledMap.put("isSuperscript", isSuperscript);
      enabledMap.put("isSubscript", isSubscript);
      return enabledMap;
    }

  }

  public enum Type {
    BOLD,
    ITALIC,
    SUBSCRIPT,
    SUPERSCRIPT,
    STRIKETHROUGH,
    UNDERLINE,
    H1,
    H2,
    H3,
    H4,
    H5,
    H6,
    ORDEREDLIST,
    UNORDEREDLIST,
    JUSTIFYCENTER,
    JUSTIFYFULL,
    JUSTIFYLEFT,
    JUSTIFYRIGHT
  }

  public interface OnTextChangeListener {
    void onTextChange(String text, String html);
  }

  public interface OnTextSelectionChangeListener {
    void onTextSelect(EnabledFormatTypes enabledFormatTypes, String selectedText);
  }

  public interface OnDecorationStateListener {
    void onStateChangeListener(String text, List<Type> types);
  }

  public interface AfterInitialLoadListener {
    void onAfterInitialLoad(boolean isReady);
  }

  public class JavascriptInterface {
    private Context ctx;

    /** Instantiate the interface and set the context */
    JavascriptInterface(Context ctx) {
      this.ctx = ctx;
    }

    /** Methods to be called from javascript page */
    @android.webkit.JavascriptInterface
    public void textChange(String text, String html) {
      RichEditor.this.textChange(text, html);
    }

    @android.webkit.JavascriptInterface
    public void selectionChange(String enabledFormatsAsQueryString, String selectedText) {
      RichEditor.this.selectionChange(enabledFormatsAsQueryString, selectedText);
    }


  }

  private static final String SETUP_HTML = "file:///android_asset/editor.html";
  private static final String STATE_SCHEME = "re-state://";
  private boolean isReady = false;
  private int textSize = 14;
  private String text; // the text written in the editor
  private String html; // the html equivalent of the text written in the editor
  private OnTextChangeListener mTextChangeListener;
  private OnTextSelectionChangeListener mTextSelectionChangeListener;
  private AfterInitialLoadListener mLoadListener;

  public RichEditor(Context context) {
    this(context, null);
  }

  public RichEditor(Context context, AttributeSet attrs) {
    this(context, attrs, android.R.attr.webViewStyle);
  }

  @SuppressLint("SetJavaScriptEnabled")
  public RichEditor(Context context, AttributeSet attrs, int defStyleAttr) {
    super(context, attrs, defStyleAttr);

    setVerticalScrollBarEnabled(false);
    setHorizontalScrollBarEnabled(false);
    getSettings().setJavaScriptEnabled(true);
    setWebChromeClient(new WebChromeClient());
    setWebViewClient(createWebviewClient());
    addJavascriptInterface(new JavascriptInterface(context), "Android");
    addJavascriptInterface(new ClipboardInterface(context), "Clipboard");
    loadUrl(SETUP_HTML);
    applyAttributes(context, attrs);
  }

  protected EditorWebViewClient createWebviewClient() {
    return new EditorWebViewClient();
  }

  public void setOnTextChangeListener(OnTextChangeListener listener) {
    this.mTextChangeListener = listener;
  }

  public void setTextSize(int textSize) {
    this.textSize = textSize;
  }

  public void setOnTextSelectionChangeListener(OnTextSelectionChangeListener listener) {
    this.mTextSelectionChangeListener = listener;
  }

  public void setOnInitialLoadListener(AfterInitialLoadListener listener) {
    this.mLoadListener = listener;
  }

  private void textChange(String text, String html) {
    this.text = text;
    this.html = html;
    if (mTextChangeListener != null) {
      mTextChangeListener.onTextChange(text, html);
    }
  }

  private void selectionChange(String queryStringWithEnabledFormats, String selectedText) {
    Uri uri = Uri.parse(queryStringWithEnabledFormats); // parse the query string and retrieve its values
    boolean isBold          = uri.getBooleanQueryParameter("isBold",          false);
    boolean isItalic        = uri.getBooleanQueryParameter("isItalic",        false);
    boolean isSubscript     = uri.getBooleanQueryParameter("isSubscript",     false);
    boolean isUnderline     = uri.getBooleanQueryParameter("isUnderline",     false);
    boolean isSuperscript   = uri.getBooleanQueryParameter("isSuperscript",   false);
    boolean isStrikethrough = uri.getBooleanQueryParameter("isStrikethrough", false);
    boolean isUnorderedList = uri.getBooleanQueryParameter("isUnorderedList", false);
    boolean isOrderedList   = uri.getBooleanQueryParameter("isOrderedList",   false);

    // create an object
    EnabledFormatTypes eft = new EnabledFormatTypes(isBold, isItalic, isUnderline, isStrikethrough, isUnorderedList, isOrderedList, isSuperscript, isSubscript);
    if (mTextSelectionChangeListener != null) {
      mTextSelectionChangeListener.onTextSelect(eft, selectedText);
    }
  }

  private void applyAttributes(Context context, AttributeSet attrs) {
    final int[] attrsArray = new int[] {
        android.R.attr.gravity
    };
    TypedArray ta = context.obtainStyledAttributes(attrs, attrsArray);

    int gravity = ta.getInt(0, NO_ID);
    switch (gravity) {
      case Gravity.LEFT:
        exec("javascript:RE.setTextAlign(\"left\")");
        break;
      case Gravity.RIGHT:
        exec("javascript:RE.setTextAlign(\"right\")");
        break;
      case Gravity.TOP:
        exec("javascript:RE.setVerticalAlign(\"top\")");
        break;
      case Gravity.BOTTOM:
        exec("javascript:RE.setVerticalAlign(\"bottom\")");
        break;
      case Gravity.CENTER_VERTICAL:
        exec("javascript:RE.setVerticalAlign(\"middle\")");
        break;
      case Gravity.CENTER_HORIZONTAL:
        exec("javascript:RE.setTextAlign(\"center\")");
        break;
      case Gravity.CENTER:
        exec("javascript:RE.setVerticalAlign(\"middle\")");
        exec("javascript:RE.setTextAlign(\"center\")");
        break;
    }

    ta.recycle();
  }

  public void setHtml(String contents) {
    if (contents == null) {
      contents = "";
    }
    try {
      exec("javascript:RE.setHtml('" + URLEncoder.encode(contents, "UTF-8") + "');");
    } catch (UnsupportedEncodingException e) {
      // No handling
    }
    this.html = contents;
  }

  public String getHtml() {
    return this.html;
  }

  public String getText() {
    return this.text;
  }

  public void setEditorFontColor(int color) {
    String hex = convertHexColorString(color);
    exec("javascript:RE.setBaseTextColor('" + hex + "');");
  }

  public void setEditorFontSize(int px) {
    exec("javascript:RE.setBaseFontSize('" + px + "px');");
  }

  @Override public void setPadding(int left, int top, int right, int bottom) {
    super.setPadding(left, top, right, bottom);
    exec("javascript:RE.setPadding('" + left + "px', '" + top + "px', '" + right + "px', '" + bottom
        + "px');");
  }

  @Override public void setPaddingRelative(int start, int top, int end, int bottom) {
    // still not support RTL.
    setPadding(start, top, end, bottom);
  }

  public void setEditorBackgroundColor(int color) {
    setBackgroundColor(color);
  }

  @Override public void setBackgroundColor(int color) {
    super.setBackgroundColor(color);
  }

  @Override public void setBackgroundResource(int resid) {
    Bitmap bitmap = Utils.decodeResource(getContext(), resid);
    String base64 = Utils.toBase64(bitmap);
    bitmap.recycle();

    exec("javascript:RE.setBackgroundImage('url(data:image/png;base64," + base64 + ")');");
  }

  @Override public void setBackground(Drawable background) {
    Bitmap bitmap = Utils.toBitmap(background);
    String base64 = Utils.toBase64(bitmap);
    bitmap.recycle();

    exec("javascript:RE.setBackgroundImage('url(data:image/png;base64," + base64 + ")');");
  }

  public void setBackground(String url) {
    exec("javascript:RE.setBackgroundImage('url(" + url + ")');");
  }

  public void setEditorWidth(int px) {
    exec("javascript:RE.setWidth('" + px + "px');");
  }

  public void setEditorHeight(int px) {
    exec("javascript:RE.setHeight('" + px + "px');");
  }

  public void setPlaceholder(String placeholder) {
    exec("javascript:RE.setPlaceholder('" + placeholder + "');");
  }

  public void setInputEnabled(Boolean inputEnabled) {
    exec("javascript:RE.setInputEnabled(" + inputEnabled + ")");
  }

  public void loadCSS(String cssFile) {
    String jsCSSImport = "(function() {" +
        "    var head  = document.getElementsByTagName(\"head\")[0];" +
        "    var link  = document.createElement(\"link\");" +
        "    link.rel  = \"stylesheet\";" +
        "    link.type = \"text/css\";" +
        "    link.href = \"" + cssFile + "\";" +
        "    link.media = \"all\";" +
        "    head.appendChild(link);" +
        "}) ();";
    exec("javascript:" + jsCSSImport + "");
  }

  public void undo() {
    exec("javascript:RE.undo();");
  }

  public void redo() {
    exec("javascript:RE.redo();");
  }

  public void setBold() {
    exec("javascript:RE.setBold();");
  }

  public void setItalic() {
    exec("javascript:RE.setItalic();");
  }

  public void setSubscript() {
    exec("javascript:RE.setSubscript();");
  }

  public void setSuperscript() {
    exec("javascript:RE.setSuperscript();");
  }

  public void setStrikeThrough() {
    exec("javascript:RE.setStrikeThrough();");
  }

  public void setUnderline() {
    exec("javascript:RE.setUnderline();");
  }

  public void setTextColor(int color) {
    exec("javascript:RE.prepareInsert();");

    String hex = convertHexColorString(color);
    exec("javascript:RE.setTextColor('" + hex + "');");
  }

  public void setTextBackgroundColor(int color) {
    exec("javascript:RE.prepareInsert();");

    String hex = convertHexColorString(color);
    exec("javascript:RE.setTextBackgroundColor('" + hex + "');");
  }

  public void setFontSize(int fontSize) {
    if (fontSize > 7 || fontSize < 1) {
      Log.e("RichEditor", "Font size should have a value between 1-7");
    }
    exec("javascript:RE.setFontSize('" + fontSize + "');");
  }

  public void removeFormat() {
    exec("javascript:RE.removeFormat();");
  }

  public void setHeading(int heading) {
    exec("javascript:RE.setHeading('" + heading + "');");
  }

  public void setIndent() {
    exec("javascript:RE.setIndent();");
  }

  public void setOutdent() {
    exec("javascript:RE.setOutdent();");
  }

  public void setAlignLeft() {
    exec("javascript:RE.setJustifyLeft();");
  }

  public void setAlignCenter() {
    exec("javascript:RE.setJustifyCenter();");
  }

  public void setAlignRight() {
    exec("javascript:RE.setJustifyRight();");
  }

  public void setBlockquote() {
    exec("javascript:RE.setBlockquote();");
  }

  public void setBullets() {
    exec("javascript:RE.setBullets();");
  }

  public void setNumbers() {
    exec("javascript:RE.setNumbers();");
  }

  public void insertImage(String url, String alt) {
    exec("javascript:RE.prepareInsert();");
    exec("javascript:RE.insertImage('" + url + "', '" + alt + "');");
  }

  public void insertLink(String href, String title) {
    exec("javascript:RE.prepareInsert();");
    exec("javascript:RE.insertLink('" + href + "', '" + title + "');");
  }

  public void insertTodo() {
    exec("javascript:RE.prepareInsert();");
    exec("javascript:RE.setTodo('" + Utils.getCurrentTime() + "');");
  }

  public void focusEditor() {
    requestFocus();
    exec("javascript:RE.focus();");
  }

  public void clearFocusEditor() {
    exec("javascript:RE.blurFocus();");
  }

  private String convertHexColorString(int color) {
    return String.format("#%06X", (0xFFFFFF & color));
  }

  protected void exec(final String trigger) {
    if (isReady) {
      load(trigger);
    } else {
      postDelayed(new Runnable() {
        @Override public void run() {
          exec(trigger);
        }
      }, 100);
    }
  }

  private void load(String trigger) {
    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT) {
      evaluateJavascript(trigger, null);
    } else {
      loadUrl(trigger);
    }
  }

  protected class EditorWebViewClient extends WebViewClient {
    @Override public void onPageFinished(WebView view, String url) {
      isReady = url.equalsIgnoreCase(SETUP_HTML);
      if (mLoadListener != null) {
        mLoadListener.onAfterInitialLoad(isReady);
      }
    }

    @Override
    public boolean shouldOverrideUrlLoading(WebView view, String url) {
      String decode;
      try {
        decode = URLDecoder.decode(url, "UTF-8");
      } catch (UnsupportedEncodingException e) {
        // No handling
        return false;
      }

      if (TextUtils.indexOf(url, STATE_SCHEME) == 0) {
        return true;
      }

      return super.shouldOverrideUrlLoading(view, url);
    }
  }
}

File: richeditor/src/main/assets/rich_editor.js
/**
 * Copyright (C) 2017 Wasabeef
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/* jshint esversion: 6 */

/* Helper Function */
// https://stackoverflow.com/questions/111529/how-to-create-query-parameters-in-javascript
function encodeQueryData(data) {
  const ret = [];
  for (var d in data) {
    if ({}.hasOwnProperty.call(data, d)) {
      ret.push(encodeURIComponent(d) + '=' + encodeURIComponent(data[d]));
    }
  }
  return "?" + ret.join('&');
}

const RE = {};

RE.currentSelection = {
  startContainer: 0,
  startOffset: 0,
  endContainer: 0,
  endOffset: 0,
};

RE.editor = document.getElementById('editor');

document.addEventListener('selectionchange', function addEventListener() {
  RE.backuprange();
  RE.selectionChange();
});


// Initializations
RE.textChange = function textChange() {
  Android.textChange(RE.getText(), RE.getHtml());
};

RE.selectionChange = function selectionChange() {
  // query all the commands for this range.
  const isBold = document.queryCommandState('bold');
  const isItalic = document.queryCommandState('italic');
  const isUnderline = document.queryCommandState('underline');
  const isStrikethrough = document.queryCommandState('strikeThrough');
  const isUnorderedList = document.queryCommandState('insertUnorderedList');
  const isOrderedList = document.queryCommandState('insertOrderedList');

    // For some reason queryCommandState is returning false for these two commands.
    //    const isSubscript     = document.queryCommandState("subscript");
    //    const isSuperscript   = document.queryCommandState("superscript");

  // So instead, i am using these alternative.
    var selection = window.getSelection();
    var nodes = [];
    var parentElement = selection.getRangeAt(0).startContainer.parentElement;
    nodes.push(parentElement); // grab an array of all nodes of the curent selection
    while(parentElement.parentNode) {
      nodes.unshift(parentElement.parentNode);
      parentElement = parentElement.parentNode;
    }

    const lowerCasedTagNamesOfNodes = nodes.map(function(el) {
      var tagName = el.tagName;
      if (tagName !== undefined) {
          return tagName.toLowerCase();
      }
    });

    const isSuperscript = lowerCasedTagNamesOfNodes.includes("sup");
    const isSubscript   = lowerCasedTagNamesOfNodes.includes("sub");

    const enabledFormatTypes = {
        isBold:          isBold,
        isItalic:        isItalic,
        isUnderline:     isUnderline,
        isStrikethrough: isStrikethrough,
        isSuperscript:   isSuperscript,
        isSubscript:     isSubscript,
        isUnorderedList: isUnorderedList,
        isOrderedList:   isOrderedList
    }
    const querystring = encodeQueryData(enabledFormatTypes);
    Android.selectionChange(querystring, RE.getText());
};

RE.setHtml = function setHtml(contents) {
  RE.editor.innerHTML = decodeURIComponent(contents.replace(/\+/g, '%20'));
};

RE.getHtml = function getHtml() {
  // removes unnecessary extra <br> tags on the beggining and end of the html string.
  const trimmedHTML = RE.editor.innerHTML.replace(/^( |<br>)*(.*?)( |<br>)*$/, '$2');
  return trimmedHTML;
};

RE.getText = function getText() {
  // trim to remove space on start and end of the text
  return RE.editor.innerText.trim();
};

RE.setBaseTextColor = function setBaseTextColor(color) {
  RE.editor.style.color = color;
};

RE.setBaseFontSize = function setBaseFontSize(size) {
  RE.editor.style.fontSize = size;
};

RE.setPadding = function setPadding(left, top, right, bottom) {
  RE.editor.style.paddingLeft = left;
  RE.editor.style.paddingTop = top;
  RE.editor.style.paddingRight = right;
  RE.editor.style.paddingBottom = bottom;
};

RE.setBackgroundColor = function setBackgroundColor(color) {
  document.body.style.backgroundColor = color;
};

RE.setBackgroundImage = function setBackgroundImage(image) {
  RE.editor.style.backgroundImage = image;
};

RE.setWidth = function setWidth(size) {
  RE.editor.style.minWidth = size;
};

RE.setHeight = function setHeight(size) {
  RE.editor.style.height = size;
};

RE.setTextAlign = function setTextAlign(align) {
  RE.editor.style.textAlign = align;
};

RE.setVerticalAlign = function setVerticalAlign(align) {
  RE.editor.style.verticalAlign = align;
};

RE.setPlaceholder = function setPlaceholder(placeholder) {
  RE.editor.setAttribute('placeholder', placeholder);
};

RE.setInputEnabled = function setInputEnabled(inputEnabled) {
  RE.editor.contentEditable = String(inputEnabled);
};

RE.undo = function undo() {
  document.execCommand('undo', false, null);
};

RE.redo = function redo() {
  document.execCommand('redo', false, null);
};

RE.setBold = function setBold() {
  document.execCommand('bold', false, null);
};

RE.setItalic = function setItalic() {
  document.execCommand('italic', false, null);
};

RE.setSubscript = function setSubscript() {
  document.execCommand('subscript', false, null);
};

RE.setSuperscript = function setSuperscript() {
  document.execCommand('superscript', false, null);
};

RE.setStrikeThrough = function setStrikeThrough() {
  document.execCommand('strikeThrough', false, null);
};

RE.setUnderline = function setUnderline() {
  document.execCommand('underline', false, null);
};

RE.setBullets = function setBullets() {
  document.execCommand('insertUnorderedList', false, null);
};

RE.setNumbers = function setNumbers() {
  document.execCommand('insertOrderedList', false, null);
};

RE.setTextColor = function setTextColor(color) {
  RE.restorerange();
  document.execCommand('styleWithCSS', null, true);
  document.execCommand('foreColor', false, color);
  document.execCommand('styleWithCSS', null, false);
};

RE.setTextBackgroundColor = function setTextBackgroundColor(color) {
  RE.restorerange();
  document.execCommand('styleWithCSS', null, true);
  document.execCommand('hiliteColor', false, color);
  document.execCommand('styleWithCSS', null, false);
};

RE.setFontSize = function setFontSize(fontSize) {
  document.execCommand('fontSize', false, fontSize);
};

RE.setHeading = function setHeading(heading) {
  document.execCommand('formatBlock', false, '<h'+heading+'>');
};

RE.setIndent = function setIndent() {
  document.execCommand('indent', false, null);
};

RE.setOutdent = function setOutdent() {
  document.execCommand('outdent', false, null);
};

RE.setJustifyLeft = function setJustifyLeft() {
  document.execCommand('justifyLeft', false, null);
};

RE.setJustifyCenter = function setJustifyCenter() {
  document.execCommand('justifyCenter', false, null);
};

RE.setJustifyRight = function setJustifyRight() {
  document.execCommand('justifyRight', false, null);
};

RE.setBlockquote = function setBlockquote() {
  document.execCommand('formatBlock', false, '<blockquote>');
};

RE.insertImage = function insertImage(url, alt) {
  var html = '<img src="' + url + '" alt="' + alt + '" />';
  RE.insertHTML(html);
};

RE.insertHTML = function insertHTML(html) {
  RE.restorerange();
  document.execCommand('insertHTML', false, html);
};

RE.insertLink = function insertLink(url, title) {
  RE.restorerange();
  const sel = document.getSelection();
  if (sel.toString().length === 0) {
    document.execCommand("insertHTML",false,"<a href='"+url+"'>"+title+"</a>");
  } else if (sel.rangeCount) {
    const el = document.createElement('a');
    el.setAttribute('href', url);
    el.setAttribute('title', title);

    const range = sel.getRangeAt(0).cloneRange();
    range.surroundContents(el);
    sel.removeAllRanges();
    sel.addRange(range);
  }
  RE.textChange();
};

RE.setTodo = function setTodo(text) {
  var html = '<input type="checkbox" name="'+ text +'" value="'+ text +'"/> &nbsp;';
  document.execCommand('insertHTML', false, html);
};

RE.prepareInsert = function prepareInsert() {
  RE.backuprange();
};

RE.backuprange = function backuprange() {
  const selection = window.getSelection();
  if (selection.rangeCount > 0) {
    const range = selection.getRangeAt(0);
    RE.currentSelection = {
      startContainer: range.startContainer,
      startOffset: range.startOffset,
      endContainer: range.endContainer,
      endOffset: range.endOffset,
    };
  }
};

RE.restorerange = function restorerange() {
  const selection = window.getSelection();
  selection.removeAllRanges();
  const range = document.createRange();
  range.setStart(RE.currentSelection.startContainer, RE.currentSelection.startOffset);
  range.setEnd(RE.currentSelection.endContainer, RE.currentSelection.endOffset);
  selection.addRange(range);
};

RE.enabledEditingItems = function enabledEditingItems(e) {
  const items = [];
  if (document.queryCommandState('bold')) {
    items.push('bold');
  }
  if (document.queryCommandState('italic')) {
    items.push('italic');
  }
  if (document.queryCommandState('subscript')) {
    items.push('subscript');
  }
  if (document.queryCommandState('superscript')) {
    items.push('superscript');
  }
  if (document.queryCommandState('strikeThrough')) {
    items.push('strikeThrough');
  }
  if (document.queryCommandState('underline')) {
    items.push('underline');
  }
  if (document.queryCommandState('orderedList')) {
    items.push('orderedList');
  }
  if (document.queryCommandState('unorderedList')) {
    items.push('unorderedList');
  }
  if (document.queryCommandState('justifyCenter')) {
    items.push('justifyCenter');
  }
  if (document.queryCommandState('justifyFull')) {
    items.push('justifyFull');
  }
  if (document.queryCommandState('justifyLeft')) {
    items.push('justifyLeft');
  }
  if (document.queryCommandState('justifyRight')) {
    items.push('justifyRight');
  }
  if (document.queryCommandState('insertHorizontalRule')) {
    items.push('horizontalRule');
  }
  const formatBlock = document.queryCommandValue('formatBlock');
  if (formatBlock.length > 0) {
    items.push(formatBlock);
  }

  window.location.href = "re-state://" + encodeURI(items.join(','));
};

RE.focus = function focus() {
  const range = document.createRange();
  range.selectNodeContents(RE.editor);
  range.collapse(false);
  const selection = window.getSelection();
  selection.removeAllRanges();
  selection.addRange(range);
  RE.editor.focus();
};

RE.blurFocus = function blurFocus() {
  RE.editor.blur();
};

RE.removeFormat = function removeFormat() {
  document.execCommand('removeFormat', false, null);
};

// Event Listeners
RE.editor.addEventListener('input', RE.textChange);
RE.editor.addEventListener('keyup', function addEventListener(e) {
  const KEY_LEFT = 37; const
    KEY_RIGHT = 39;
  if (e.which === KEY_LEFT || e.which === KEY_RIGHT) {
    RE.enabledEditingItems(e);
  }
});
RE.editor.addEventListener('click', RE.enabledEditingItems);
RE.editor.addEventListener('paste', function addEventListener(evt) {
  evt.preventDefault();
  const plain = Clipboard.getText();
  RE.insertHTML(plain);
});
